table/tables/cache.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tables
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/tikv/client-go/v2/tikv"
+)
+
+var _ table.Table = &cachedTable{}
+var _ table.CachedTable = &cachedTable{}
+
+type cachedTable struct {
+	TableCommon
+	kv.MemBuffer
+}
+
+func (c *cachedTable) IsReadFromCache(ts uint64) bool {
+	// If first read cache table. directly return false, the backend goroutine will help us update the lock information
+	// and read the data from the original table at the same time
+	// TODO : Use lease and ts judge whether it is readable.
+	// TODO : If the cache is not readable. MemBuffer become invalid.
+	return c.MemBuffer != nil
+}
+
+func (c *cachedTable) GetMemCache() kv.MemBuffer {
+	return c.MemBuffer
+}
+
+// NewCachedTable creates a new CachedTable Instance
+func NewCachedTable(tbl *TableCommon) (table.Table, error) {
+	return &cachedTable{
+		TableCommon: *tbl,
+	}, nil
+}
+
+func (c *cachedTable) LoadDataFromOriginalTable(ctx sessionctx.Context) error {
+	prefix := tablecodec.GenTablePrefix(c.tableID)
+	txn, err := ctx.Txn(true)
+	if err != nil {
+		return err
+	}
+	buffTxn, err := ctx.GetStore().BeginWithOption(tikv.DefaultStartTSOption().SetStartTS(0))
+	if err != nil {
+		return err
+	}
+	c.MemBuffer = buffTxn.GetMemBuffer()
```

Comment:
We need to protect access of `c.MemBuffer` to avoid data race. 

1. assign MemBuffer to cacheTable after data loaded instead of it is created
2. Use `sync.RWMutex` to protect set and get

Reply:
Ok understand